pkg/metrics/status.go

```diff
@@ -154,4 +195,98 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
+			continue
+		}
+
+		targetClusterName, targetNodeName := getClusterNodeName(n.Name)
+		nodePathStatus := healthClientPkg.GetHostPrimaryAddress(n)
+		nodePathConnectivityStatusType := healthClientPkg.GetPathConnectivityStatusType(nodePathStatus)
+		endpointPathStatus := n.HealthEndpoint
+		endpointPathConnectivityStatusType := healthClientPkg.GetPathConnectivityStatusType(endpointPathStatus.PrimaryAddress)
+		isEndpointReachable := endpointPathConnectivityStatusType == healthClientPkg.ConnStatusReachable
+		isNodeReachable := nodePathConnectivityStatusType == healthClientPkg.ConnStatusReachable
+		nodeType := LabelValueLocal
+
+		if targetClusterName != localClusterName {
+			nodeType = LabelValueRemote
+		}
+
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityStatusDesc,
+			prometheus.GaugeValue,
+			boolToFloat64(isEndpointReachable),
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueEndpoint,
+		)
+
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityStatusDesc,
+			prometheus.GaugeValue,
+			boolToFloat64(isNodeReachable),
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueNode,
+		)
+
+		var metricValue float64
+
+		metricValue = -1
+		if endpointPathStatus.PrimaryAddress.HTTP != nil {
+			metricValue = float64(endpointPathStatus.PrimaryAddress.HTTP.Latency) / float64(time.Second)
+		}
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityLatencyDesc,
+			prometheus.GaugeValue,
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueEndpoint, LabelValueHTTP,
+		)
+
+		metricValue = -1
+		if nodePathStatus.HTTP != nil {
+			metricValue = float64(nodePathStatus.HTTP.Latency) / float64(time.Second)
+		}
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityLatencyDesc,
+			prometheus.GaugeValue,
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueNode, LabelValueHTTP,
+		)
+
+		metricValue = -1
+		if endpointPathStatus.PrimaryAddress != nil && endpointPathStatus.PrimaryAddress.Icmp != nil {
+			metricValue = float64(endpointPathStatus.PrimaryAddress.Icmp.Latency) / float64(time.Second)
+		}
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityLatencyDesc,
+			prometheus.GaugeValue,
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueEndpoint, LabelValueICMP,
+		)
+
+		metricValue = -1
+		if nodePathStatus.Icmp != nil {
+			metricValue = float64(nodePathStatus.Icmp.Latency) / float64(time.Second)
+		}
+
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityLatencyDesc,
+			prometheus.GaugeValue,
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueNode, LabelValueICMP,
+		)
+	}
```

Comment:
That is a very good question, but I don't have a good answer. 

Overall, I do think we *should report on the state of the secondary address*, as any errors on the secondary addresses will likely manifest as connectivity problems with IPv6 in dual-stack setups. The downstream metrics likely didn't consider them because the secondary address status was not exposed in the API until recently.

For the `status` (i.e. reachable vs non-reachable) I think we can consider the secondary paths easily by summarizing the status of all paths. Meaning: If one of the paths (e.g. secondary address via ICMP) to the node or endpoint is unhealthy, we report the whole target (node or endpoint) as unhealthy. This is what we already do for the existing metrics and brief output (see code below), and I think is consistent with the usage of `GetPathConnectivityStatusType` in the downstream community code which also summarizes HTTP and ICMP together.  After all, for full connectivity we need all paths to be healthy.

https://github.com/cilium/cilium/blob/8e25d64386426ea97fa7b3a35f77d074ef4a7546/pkg/health/client/client.go#L316-L317
https://github.com/cilium/cilium/blob/88d16c36037fb50e6838d849ff930ea0451ac711/pkg/metrics/status.go#L130-L144

For `latency` however, the story is more complicated. Here I think it's mathematically wrong to try to summarize and take the average of the latency of all paths, since each paths will have its own latency characteristics/distribution and therefore the average will pretty much be meaningless.

But by collecting the latency for each path separately will double our metrics, as we will have 8 latency values (`(node/endpoint) x (primary/secondary) x (icmp/http)` for each node to node pair instead of the existing 4.

Reply:
Thanks both for your inputs. 

I can make update to aggregate status and expose value accordingly.

For latency, further to Sebastian's comment, SecondaryAddresses is a slice, so the total number of metrics per latency could potentially be even more. We can add feature gate like what we have right now (i.e. EnableNodeConnectivityLatencyMetric). Keen to hear your input here.

Reply:
The way `SecondaryAddresses` is implemented right now, it will contain none or at most one entry. I don't expect to change it any time soon, but good to keep in mind nonetheless.

Yeah, I would love for the metrics to be more configurable. For the Hubble metrics, one can configure each metric separately: https://docs.cilium.io/en/v1.11/operations/metrics/#hubble-metrics

Might be worth evaluating if that approach could also work for the normal Cilium metrics, instead of adding a ton of feature flags.

Reply:
The mechanism is a little bit different in cilium metrics, but functionality is there as long as we have these two new metrics in default list.

https://github.com/cilium/cilium/blob/master/Documentation/cmdref/cilium-agent.md
```
      --metrics strings                                      Metrics that should be enabled or disabled from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```

Reply:
So in Hubble, I think the mechanism is even a bit more powerful: You can set options per metrics via this flag too. This would allow us to e.g. have something like `--metrics +health_status:include_secondary,+health_latency:include_secondary` to e.g. enabling the secondary addresses for these metrics. But I have not yet looked at how these configuration flags are parsed.

Reply:
thanks, I can spin off the above metric configuration to another github issue if you don't mind.

Reply:
Sounds good to me!